src/core/mac/mac_frame.cpp

```diff
@@ -1165,16 +1165,20 @@ otError TxFrame::GenerateEnhAck(const RxFrame &aFrame, bool aIsFramePending, con
         SetDstAddr(address);
     }
 
-    // Set security header
-    SuccessOrExit(error = aFrame.GetSecurityControlField(securityControlField));
-    SuccessOrExit(error = aFrame.GetFrameCounter(frameCounter));
-    SuccessOrExit(error = aFrame.GetKeyId(keyId));
-
     SetPsduLength(kMaxPsduSize); // At this time the length of ACK hasn't been determined, set it to
                                  // `kMaxPsduSize` to call methods that check frame length.
-    SetSecurityControlField(securityControlField);
-    SetFrameCounter(frameCounter);
-    SetKeyId(keyId);
+
+    // Set security header
+    if (aFrame.GetSecurityEnabled())
```

Comment:
Per 802.15.4-2015 6.7.2

> If the Enh-Ack frame contains IEs and/or a Frame Payload and it is in responseto a secured frame, then the Enh-Ack frame shall be secured. If the Enh-Ack frame is in response to asecured frame and does not contain either IEs or a Frame Payload, then the Enh-Ack may be secured.

So only when Rx frame is a secured frame and the enh-ack contains IEs/Payloads, it needs to be secured. This code only check for Rx frame security. Are there any modification in Thread spec to specify this behavior? I.E, enh-ack should have same security with regard to the received frame?

Reply:
Thread spec doesn't specify this. Actually I implemented this referring to the enh ack generation implemented in nrf radio driver. But I think it doesn't violate the spec to always secure the Enh-Ack when the Rx frame is secured no matter if the Enh-Ack contains IE or Payload. Will this cause any other problems?

Reply:
From the 15.4 spec, I think `may be secured` means we can always use secured enh-ACK for received secure 15.4-2015 frame. But I'm considering something looking paradoxical to me.

15.4-2015 specify that retransmission should be using the exactly the same frame

> 6.7.4.3 When not using TSCH mode and a frame with the Security Enabled field set to one is retransmitted, the frame shall be retransmitted without changes and without passing through the outgoing frame security procedure, as defined in 9.2.1.

However, there are cases when this limitation does not hold true. For example, the CSL IE should be updated so that the CSL phase is correct.

Another case is, when doing retransmission, there are backoff between retransmission. What if peer send a secured 2015 frame to this node and this node response with an Enh-ACK with auxiliary frame counter increased. Then the old frame counter in the retransmission is not valid any more?

Reply:
As discussed. It is OK to always secure enh-ack for received secure 15.4-2015 frame. Set to resolved.